packages/core/src/util/security/trusted_types.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+/**
+ * A class to facilitate use of Trusted Types policies within Angular. Lazily
+ * constructs a Trusted Types policy with the specified name, providing helper
+ * utilities for promoting strings to Trusted Types. When Trusted Types are not
+ * available, strings are used as a fallback.
+ */
+export class AngularTrustedTypesPolicy {
+  /**
+   * The wrapped Trusted Types policy, or null if Trusted Types are
+   * not enabled/supported, or undefined if the policy has not been created yet.
+   */
+  private policy: TrustedTypePolicy|null|undefined;
+
+  constructor(private policyName: string) {}
+
+  /**
+   * Returns the wrapped Trusted Types policy, or null if Trusted Types are not
+   * enabled/supported. The first call to this function will create the policy.
+   */
+  private getPolicy(): TrustedTypePolicy|null {
+    if (this.policy === undefined) {
+      this.policy = null;
+      if (typeof window !== 'undefined' && window.trustedTypes) {
+        try {
+          this.policy = window.trustedTypes.createPolicy(this.policyName, {
+            createHTML: (s: string) => s,
+            createScript: (s: string) => s,
+            createScriptURL: (s: string) => s,
+          });
+        } catch {
+          // trustedTypes.createPolicy throws if called with a name that is
+          // already registered, even in report-only mode. Until the API changes,
+          // catch the error not to break the applications functionally. In such
+          // cases, the code will fall back to using strings.
+        }
+      }
+    }
+    return this.policy;
+  }
+
+  /**
+   * Unsafely promote a string to a TrustedHTML, falling back to strings when
+   * Trusted Types are not available. This is a security-sensitive function; any
+   * use of this function must go through security review. In particular, it must
+   * be assured that the provided string will never cause an XSS vulnerability if
+   * used in a context that will be interpreted as HTML by a browser, e.g. when
+   * assigning to element.innerHTML.
+   */
+  trustedHTMLFromStringRequiresSecurityReview(html: string): TrustedHTML|string {
+    return this.getPolicy()?.createHTML(html) || html;
```

Comment:
It doesn't matter with the current implementation, but I believe using `||` is unsafe here. For example, if the policy is updated in the future to strip off certain tags.

Imagine the following:

```ts
this.policy = window.trustedTypes.createPolicy(this.policyName, {
  // DISCLAIMER: For demonstration purposes only.
  createHTML: (s: string) => s.replace(/<script>[\s\S]*?<\/script>/g, ''),
  ...
});
```

With the above policy, `trustedHTMLFromStringRequiresSecurityReview('<script>alert("Hax")</script>')` would return the original, unsafe string, even though we have a policy that is supposed to strip inline scripts.

Maybe you could instead use:

```suggestion
    return this.getPolicy()?.createHTML(html) ?? html;
```

##
The same goes for the other `trustedXyzFromStringRequiresSecurityReview()` methods.

Reply:
Alternatively, you could update `getPolicy()` to always return a "policy", which in the absense of Trusted Types would be:

```ts
{
  createHTML: (s: string) => s,
  createScript: (s: string) => s,
  createScriptURL: (s: string) => s,
}
```

Reply:
It's a good observation, but misses one fact: a Trusted Type (even if it wraps the empty string) is always a thruthy value. So if Trusted Types are available at all, that expression will always evaluate to the Trusted Type on the left. I intentionally used `||` here as a minor optimization, since `??` compiles down to a slightly larger expression. I'm fine with switching to `??` if you still have concerns.

I like the idea of using the placeholder policy when Trusted Types are not available, but suspect the necessary changes to the type definition of the policy variable will cancel out the elegance that it brings.

Reply:
Ah, good point 😇 In that case, `||` is fine 👍

Yeah, defining a placeholder policy might get messy or bloaty. Let's keep it as is.